# Production-Grade House Price Prediction ML Pipeline
### An End-to-End Educational & Production-Quality Case Study
This notebook walks through a complete Machine Learning workflow, from business understanding to model serialization, following industry best practices in Data Science and Machine Learning Engineering.


## Phase 1 — Business Understanding

### 1. Objective
To understand the business context of house price prediction, map it to a machine learning task, define success criteria, and establish metrics.

### 2. Theory
Real estate valuation is a critical process for buyers, sellers, lenders, and online portals (e.g., Zillow, Opendoor). Accurate pricing reduces friction in transactions, optimizes mortgage lending portfolios, and provides transparency to the market.

### 3. Why this step is necessary
Without a clear business objective, machine learning projects often suffer from misalignment (e.g., optimizing for a metric that doesn't capture business value or failing to account for real-world deployment limitations).

### 4. Mathematical Intuition
We define the target variable $y_i$ as the actual Sale Price and $\hat{y}_i$ as the predicted Sale Price.
The primary business metric is Mean Absolute Error (MAE):
$$MAE = \frac{1}{N} \sum_{i=1}^N |y_i - \hat{y}_i|$$
And Root Mean Squared Error (RMSE) on log-transformed prices to evaluate percentage errors:
$$RMSE_{\log} = \sqrt{\frac{1}{N} \sum_{i=1}^N (\log(y_i + 1) - \log(\hat{y}_i + 1))^2}$$

### 5. Python Implementation


In [1]:
# Define baseline metrics for business goals
TARGET_RMSE_LOG = 0.13  # Representing ~13% error
TARGET_MAE_USD = 15000  # $15,000 USD average error
print(f"Target Performance Goals:\n- RMSE Log: {TARGET_RMSE_LOG}\n- MAE USD: ${TARGET_MAE_USD}")



Target Performance Goals:
- RMSE Log: 0.13
- MAE USD: $15000


### 6. Explanation of the Code
We establish constants for our target model performance based on business constraints.

### 7. Interpretation of the Output
The baseline target is defined: we want a model that predicts house values within a ~13% log-margin, equivalent to approximately $15,000 on average for mid-range homes.

### 8. Common Mistakes
- Relying purely on R-squared which scales with the variance of the test set, instead of scale-dependent metrics like MAE or RMSE.
- Failing to account for log-scale evaluation, which treats a $10k error on a $100k home and a $100k error on a $1M home with equal relative weight.

### 9. Best Practices
- Define KPIs alongside stakeholders before building models.
- Set up a log-evaluation metric for housing price prediction since prices are right-skewed and errors should be relative.




## Phase 2 — Data Understanding

### 1. Objective
To load, inspect, and analyze the dataset dimensions, types, missing values, duplicates, and target variable characteristics.

### 2. Theory
Data understanding involves examining the schema, rows, columns, and properties of the files. The Ames Housing dataset has 1,460 samples (train) and 81 columns, including 79 explanatory features, 1 ID, and 1 Target variable (`SalePrice`).

### 3. Why this step is necessary
It allows us to check for corrupted files, understand types (nominal, ordinal, continuous, discrete), and plan cleaning/preprocessing pipelines.

### 4. Mathematical Intuition
We evaluate the distribution shape of the target variable `SalePrice` using Skewness ($S$) and Kurtosis ($K$):
$$S = \frac{\frac{1}{N} \sum_{i=1}^N (y_i - \bar{y})^3}{s^3}$$
$$K = \frac{\frac{1}{N} \sum_{i=1}^N (y_i - \bar{y})^4}{s^4} - 3$$
Where $\bar{y}$ is the sample mean, and $s$ is the standard deviation. A skewness of 0 represents a perfectly symmetric distribution; kurtosis measures the tail thickness.

### 5. Python Implementation


In [2]:
import pandas as pd
import numpy as np
import os

# Define file paths
train_path = "../data/train.csv"
test_path = "../data/test.csv"

# Load datasets
df_train = pd.read_csv(train_path)
df_test = pd.read_csv(test_path)

# 1. Shape
print(f"Train Shape: {df_train.shape}")
print(f"Test Shape: {df_test.shape}")

# 2. Head & Tail
print("\nFirst 3 rows of train set:")
display(df_train.head(3))

print("\nLast 3 rows of train set:")
display(df_train.tail(3))

# 3. Sample rows
print("\nRandom Sample of 3 rows:")
display(df_train.sample(3, random_state=42))

# 4. Info & Data Types
print("\nData Types & Info:")
df_train.info()

# 5. Summary Statistics
print("\nSummary Statistics of Numeric Columns:")
display(df_train.describe().T.head(10))

# 6. Missing Values
missing_val = df_train.isnull().sum()
missing_pct = 100 * missing_val / len(df_train)
missing_df = pd.DataFrame({"Missing Count": missing_val, "Percentage (%)": missing_pct})
missing_df = missing_df[missing_df["Missing Count"] > 0].sort_values(by="Missing Count", ascending=False)
print("\nColumns with Missing Values:")
display(missing_df.head(10))

# 7. Duplicate Rows
duplicates = df_train.duplicated().sum()
print(f"\nDuplicate rows found: {duplicates}")

# 8. Cardinality of Categorical Columns
cat_cols = df_train.select_dtypes(include=["object"]).columns
cardinality = df_train[cat_cols].nunique().sort_values(ascending=False)
print("\nCardinality of Categorical Columns:")
display(cardinality.head(10))

# 9. Target Variable Analysis
skew = df_train["SalePrice"].skew()
kurt = df_train["SalePrice"].kurt()
print(f"\nTarget Variable ('SalePrice') Statistics:\n- Mean: ${df_train['SalePrice'].mean():.2f}\n- Median: ${df_train['SalePrice'].median():.2f}\n- Skewness: {skew:.4f}\n- Kurtosis: {kurt:.4f}")



Train Shape: (1460, 81)
Test Shape: (1459, 80)

First 3 rows of train set:


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500



Last 3 rows of train set:


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
1457,1458,70,RL,66.0,9042,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,GdPrv,Shed,2500,5,2010,WD,Normal,266500
1458,1459,20,RL,68.0,9717,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,4,2010,WD,Normal,142125
1459,1460,20,RL,75.0,9937,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,6,2008,WD,Normal,147500



Random Sample of 3 rows:


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
892,893,20,RL,70.0,8414,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,MnPrv,NaN,0,2,2006,WD,Normal,154500
1105,1106,60,RL,98.0,12256,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,4,2010,WD,Normal,325000
413,414,30,RM,56.0,8960,Pave,Grvl,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,3,2010,WD,Normal,115000



Data Types & Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   i

,count,mean,std,min,25%,50%,75%,max
Id,1460.0,730.500000,421.610009,1.0,365.75,730.5,1095.25,1460.0
MSSubClass,1460.0,56.897260,42.300571,20.0,20.00,50.0,70.00,190.0
LotFrontage,1201.0,70.049958,24.284752,21.0,59.00,69.0,80.00,313.0
LotArea,1460.0,10516.828082,9981.264932,1300.0,7553.50,9478.5,11601.50,215245.0
OverallQual,1460.0,6.099315,1.382997,1.0,5.00,6.0,7.00,10.0
OverallCond,1460.0,5.575342,1.112799,1.0,5.00,5.0,6.00,9.0
YearBuilt,1460.0,1971.267808,30.202904,1872.0,1954.00,1973.0,2000.00,2010.0
YearRemodAdd,1460.0,1984.865753,20.645407,1950.0,1967.00,1994.0,2004.00,2010.0
MasVnrArea,1452.0,103.685262,181.066207,0.0,0.00,0.0,166.00,1600.0
BsmtFinSF1,1460.0,443.639726,456.098091,0.0,0.00,383.5,712.25,5644.0



Columns with Missing Values:


,Missing Count,Percentage (%)
PoolQC,1453,99.520548
MiscFeature,1406,96.301370
Alley,1369,93.767123
Fence,1179,80.753425
MasVnrType,872,59.726027
FireplaceQu,690,47.260274
LotFrontage,259,17.739726
GarageType,81,5.547945
GarageYrBlt,81,5.547945
GarageFinish,81,5.547945



Duplicate rows found: 0

Cardinality of Categorical Columns:


Neighborhood    25
Exterior2nd     16
Exterior1st     15
Condition1       9
SaleType         9
HouseStyle       8
RoofMatl         8
Condition2       8
Functional       7
BsmtFinType2     6
dtype: int64


Target Variable ('SalePrice') Statistics:
- Mean: $180921.20
- Median: $163000.00
- Skewness: 1.8829
- Kurtosis: 6.5363


### 6. Explanation of the Code
We read the training and test CSV files using Pandas, display their structural attributes, summarize numeric distributions, calculate missing value statistics, count duplicates, and compute skewness/kurtosis of the target variable `SalePrice`.

### 7. Interpretation of the Output
- The training set contains 1,460 observations and 81 columns, while the test set has 1,459 observations and 80 columns (excluding the target column `SalePrice`).
- The dataset contains many missing values (e.g. `PoolQC`, `MiscFeature`, `Alley`, `Fence` have >80% missing data).
- The target variable `SalePrice` is highly right-skewed (skewness ≈ 1.88) and leptokurtic (kurtosis ≈ 6.53), showing that a log transformation is highly recommended.

### 8. Common Mistakes
- Neglecting to inspect the test dataset shape and content.
- Missing the fact that `MSSubClass` is numeric but actually represents a categorical variable (building class), which needs casting.

### 9. Best Practices
- Always check missing percentage values to decide whether to drop columns or impute them.
- Look at the cardinality of text categories to prepare for encoding (high cardinality, e.g. `Neighborhood` with 25 categories, requires special care).


